### PR TITLE
register callback for marshal in VM

### PR DESCRIPTION
**Pull request by ringabout**
ref  https://github.com/nim-lang/RFCs/issues/437#issuecomment-1015878460

### CODE REVIEW

File: compiler/vmdef.nim [hunk 1]
```diff
@@ -238,6 +237,7 @@ type
     slots*: ptr UncheckedArray[TFullReg]
     currentException*: PNode
     currentLineInfo*: TLineInfo
+    callbackNode*: PNode
```
Comment (Araq) [inline compiler/vmdef.nim]: Why is this required? We have dozens of callbacks already which work fine without this.

Comment (ringabout) [reply compiler/vmdef.nim]: This is the signature of `marshal.to`. I don't know how to get the type of generics T from VmArgs.
```nim
proc to*[T](data: string): T
```
which is required by `proc storeAny*(s: var string; t: PType; a: PNode; conf: ConfigRef)` (in vmmarshal)

`getString(a, 0)`, `getNode(a, 0)` etc. seem to only work for formal parameters not generic parameters or return types.

File: lib/pure/marshal.nim [hunk 2]
```diff
@@ -298,6 +298,9 @@ proc store*[T](s: Stream, data: T) =
   shallowCopy(d, data)
   storeAny(s, toAny(d), stored)
 
+proc loadVM[T](typ: typedesc[T], x: T): string =
+  discard "the implementation is in the compiler/vmops"
+
 proc `$$`*[T](x: T): string =
```
Comment (ringabout) [inline lib/pure/marshal.nim:304]: Without registering `loadVM`, it seems to crash:

```
C:\Users\blue\Desktop\work\Nim\compiler\vm.nim(2226) evalStaticStmt
C:\Users\blue\Desktop\work\Nim\compiler\vm.nim(2215) evalConstExprAux
C:\Users\blue\Desktop\work\Nim\compiler\vm.nim(1224) rawExecute
C:\Users\blue\Desktop\work\Nim\compiler\vmops.nim(383) :anonymous
C:\Users\blue\Desktop\work\Nim\compiler\vmmarshal.nim(135) storeAny
C:\Users\blue\Desktop\work\Nim\compiler\vmmarshal.nim(59) storeAny
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

**Comment by ringabout** [comment]
Now it uses two helper procs `loadVM` and `toVM` to bypass the limitations of `VmArgs` -- cannot get the types of return values or generic parameters.
